src/main/java/org/opensearch/security/resources/ResourceSharingIndexHandler.java

```diff
@@ -0,0 +1,1175 @@
+        updateScriptListener.whenComplete(success -> {
+            if (!success) {
+                LOGGER.error("Failed to update resource sharing info for resource {}", resourceId);
+                listener.onResponse(null);
+                return;
+            }
+            // TODO check if this should be replaced by Java in-memory computation (current intuition is that it will be more memory
+            // intensive to do it in java)
+            fetchResourceSharingDocument(sourceIdx, resourceId, updatedSharingListener);
+        }, listener::onFailure);
+
+        updatedSharingListener.whenComplete(listener::onResponse, listener::onFailure);
+    }
+
+    /**
+     * Revokes access for specified entities from a resource sharing document. This method removes the specified
+     * entities (users, roles, or backend roles) from the existing sharing configuration while preserving other
+     * sharing settings.
+     *
+     * <p>The method performs the following steps:
+     * <ol>
+     *   <li>Fetches the existing document</li>
+     *   <li>Removes specified entities from their respective lists in all sharing groups</li>
+     *   <li>Updates the document if modifications were made</li>
+     *   <li>Returns the updated resource sharing configuration</li>
+     * </ol>
+     *
+     * <p>Example document structure:
+     * <pre>
+     * {
+     *   "source_idx": "resource_index_name",
+     *   "resource_id": "resource_id",
+     *   "share_with": {
+     *     "action-group": {
+     *       "users": ["user1", "user2"],
+     *       "roles": ["role1", "role2"],
+     *       "backend_roles": ["backend_role1"]
+     *     }
+     *   }
+     * }
+     * </pre>
+     *
+     * @param resourceId      The ID of the resource from which to revoke access
+     * @param sourceIdx       The name of the system index where the resource exists
+     * @param revokeAccess    A map containing entity types (USER, ROLE, BACKEND_ROLE) and their corresponding
+     *                        values to be removed from the sharing configuration
+     * @param actionGroups     A set of action-groups to revoke access from. If null or empty, access is revoked from all action-groups
+     * @param requestUserName The user trying to revoke the accesses
+     * @param isAdmin         Boolean indicating whether the user is an admin or not
+     * @param listener        Listener to be notified when the operation completes
+     * @throws IllegalArgumentException if resourceId, sourceIdx is null/empty, or if revokeAccess is null/empty
+     * @throws RuntimeException         if the update operation fails or encounters an error
+     * @apiNote This method modifies the existing document. If no modifications are needed (i.e., specified
+     * entities don't exist in the current configuration), the original document is returned unchanged.
+     * @see Recipient
+     * @see ResourceSharing
+     */
+    public void revokeAccess(
```

Comment:
I don't think it makes sense to have this in the java API. From a plugin dev point of view I want to know which resources the user has access to before issuing a query for them (akin to DLS, but shifting the responsibility to the plugin dev)

Meanwhile, security can expose 2 REST APIs;

- Get sharing info of the resource (can be queried by resource_type and resource_id)
- Update sharing info - again can be called with the resource_type and resource_id

When this feature is more mature there would additionally be re-usable components that call the security API behind the scenes if security is installed and enabled.

Reply:
Since no plugins actually call revoke atm, this method was added as a working example that is available for plugins to call.

In future, we can open this by replacing share and revoke from client to REST APIs, however that will move away of current default behaviour of sharing with backend-roles automatically and will be considered breaking change for new resources being created. Before introducing this a migration path will have to be in place to allow existing resources and their sharing to be migrated to central index.